### PR TITLE
PHP8.1 support

### PR DESCRIPTION
**Pull request by SpacePossum**
@kubawerlos do you maybe have time to run the tests with the top X PHP projects to see if there is regression we overlooked?

### CODE REVIEW

File: php-cs-fixer [hunk 1]
```diff
@@ -40,9 +40,9 @@ set_error_handler(static function ($severity, $message, $file, $line) {
         exit(1);
     } elseif (
         \PHP_VERSION_ID < 70205
-        || \PHP_VERSION_ID >= 80100
+        || \PHP_VERSION_ID >= 80200
```
Comment (keradus) [inline php-cs-fixer:43]: we should update `.php-cs-fixer.php-highest.php:13:if (PHP_VERSION_ID <= 80001 || PHP_VERSION_ID >= 80100) {` too

Comment (SpacePossum) [reply php-cs-fixer:43]: thanks, updated, also addressed some other minors, please have a look again :)

File: .php-cs-fixer.php-highest.php [hunk 2]
```diff
@@ -10,16 +10,16 @@
  * with this source code in the file LICENSE.
  */
 
-if (PHP_VERSION_ID <= 80001 || PHP_VERSION_ID >= 80100) {
-    fwrite(STDERR, "PHP CS Fixer's config for PHP-HIGHEST can be executed only on highest supported PHP version - 8.0.*.\n");
+if (PHP_VERSION_ID <= 80001 || PHP_VERSION_ID >= 80200) {
```
Comment (kubawerlos) [inline .php-cs-fixer.php-highest.php:13]:
```suggestion
if (PHP_VERSION_ID < 80100 || PHP_VERSION_ID >= 80200) {
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -46,10 +49,6 @@ jobs:
             calculate-code-coverage: 'yes'
             phpunit-flags: '--testsuite coverage --exclude-group covers-nothing --coverage-clover build/logs/clover.xml'
 
-          - operating-system: 'ubuntu-20.04'
-            php-version: '8.1'
```
Comment (keradus) [inline .github/workflows/ci.yml:50]: should we add 8.2 snapshot, or not yet available?

**Review by Ocramius — CHANGES_REQUESTED** [review]
Since `php-cs-fixer` declares `^8.0` as `composer.json` supported version:

https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/4ab8dc5a52e6bdf183865cc6093780ac6774c1fa/composer.json#L17

Then crashes at runtime (message changed in this patch):

```
PHP needs to be a minimum version of PHP 7.2.5 and maximum version of PHP 8.0.*.
```

 I'd say that changing `composer.json` to restrict to `"php": "^7.2.5 || ~8.0.0 || ~8.1.0"` is a required addition to this patch: as it currently stands, the fact that the tool installs cleanly, then fails at a later point in time, is not really useful.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (rvanlaak) [reply .github/workflows/ci.yml:50]: Yes, for Github Action `shivammathur/setup-php@v2` 8.2 is available as nightly (and will automatically transition to stable once released): https://github.com/marketplace/actions/setup-php-action#tada-php-support

Including 8.1 in our matrix did work very well last month's PHP release.

**Comment by keradus** [comment]
@Ocramius we had tons of ppl crying that they cannot update their project to use new PHP because PHP CS Fixer is not yet compatible. EVERY SINGLE TIME new PHP is out.
We don't like either of those situations, yet we decided to go with one in the end - runtime check. Feel free to check previous discussions around this topic.

**Comment by Ocramius** [comment]
I tell them to wait: having a tool that you install (then crashes at startup) is really no better, and it just gives false confidence (and introduces instability in the overall ecosystem).

The fact that dependency ranges don't match actual compatibility is mostly something that **some communities** have erroneously started advocating for, because solving the compatibility is less important than having users upgrade :shrug: 

See also https://github.com/laminas/technical-steering-committee/blob/c7f6babfb1773d9cb9371208102cbb8b27980bd5/meetings/minutes/2021-08-02-TSC-Minutes.md#locked-dependencies-for-all-supported-php-versions for more context, previously discussed by the @laminas TSC.

In practice, for my own packages:

 * I pin to minor PHP: `{"require": {"php": "~7.4.0 || ~8.0.0 || ~8.1.0"}}` (**no `~8.2.0` until I've tested it in RC!!!**)
 * I tell users that they are free to shoot themselves in the foot with `composer update --ignore-platform-req=php`

**Comment by keradus** [comment]
Unfortunately while we were following approach you describe, we faced many issues.

Anyway, feel free to open a new discussion, as this is not the blocker for this PR (as we have a current practice in place for a year) and I would prefer to not hijack this thread but focus on 8.1 here.

<details>
<summary>unfortunate note</summary>

I also spent quite a time discussing all pros and cons about having ~8.0 vs ^8 vs >=8. TBH, I rather prefer to spend time ensuring PHP8.x compatibility rather than spending time in a discussion that year ago started to be similar to tabs vs spaces.
</details>

File: .php-cs-fixer.php-highest.php (hunk 2 above)
Comment (SpacePossum) [reply .php-cs-fixer.php-highest.php:13]: I think the issue on 8.0.0 was solved in 8.0.1, so we can run on anything above it

File: .github/workflows/ci.yml (hunk 3 above)
Comment (SpacePossum) [reply .github/workflows/ci.yml:50]: https://github.com/marketplace/actions/setup-php-action#nightly-build-setup

seems there is 8.2/nightly support, let me try, also nice to see the tool listed there as an example :)

File: .php-cs-fixer.php-highest.php (hunk 2 above)
Comment (keradus) [reply .php-cs-fixer.php-highest.php:13]: this check was to ensure that `php-highest` is executed only on ~8.0.1 originally, and now we should ensure it's only ~8.1.
you allow ~8.0 as well

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -10,54 +10,58 @@ jobs:
...
-
-          - operating-system: 'ubuntu-20.04'
-            php-version: '8.0'
-            job-description: 'with calculating code coverage'
-            calculate-code-coverage: 'yes'
-            phpunit-flags: '--testsuite coverage --exclude-group covers-nothing --coverage-clover build/logs/clover.xml'
-
-          - operating-system: 'ubuntu-20.04'
-            php-version: '8.1'
+            php-version: 'nightly'
```
Comment (keradus) [inline .github/workflows/ci.yml:53]: @SpacePossum , if 8.2 doesn't run properly for us, let's revert, merge this PR without 8.2 in pipeline and work on 8.2 pipeline as separated PR

**Comment by Bilge** [comment]
Waiting on this. Guess we'll have to disable PHP-CS-Fixer for now.

**Comment by vbasky** [comment]
> @Ocramius we had tons of ppl crying that they cannot update their project to use new PHP because PHP CS Fixer is not yet compatible. EVERY SINGLE TIME new PHP is out. Same time, not so many ppl are eager to help making PHP CS Fixer understanding new PHP syntax or be compatible with modified tokenizer.
> We don't like either of those situations, yet we decided to go with one in the end - runtime check. Feel free to check previous discussions around this topic.

I'd love to help but unsure where to start on the repo

**Review by kubawerlos — APPROVED** [review]
@SpacePossum finally the check finished, nothing related to PHP 8.1, only stuff that are already released (PR with fix in progress).

**Comment by SpacePossum** [comment]
Thanks so much @kubawerlos :+1: 

I think that is it, all feedback processed. If the tests are green I plan on merging. If anyone sees some related issue still open please comment :)

**Comment by keradus** [comment]
kudos for executing the regression, @kubawerlos !
@SpacePossum , we are good to go ;)

**Review by keradus — APPROVED** [review]

**Comment by iwiznia** [comment]
Sorry if I am not understanding this correctly, but it seems all [issues to support php 8.1 are done](https://github.com/FriendsOfPHP/PHP-CS-Fixer/labels/topic%2FPHP8.1) but composer.json still [does not allow 8.1](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/composer.json#L17). 
What is missing to change that to allow php 8.1?

**Comment by Ocramius** [comment]
@iwiznia `^8.0` means `>=8.0.0,<9.0.0`

**Comment by iwiznia** [comment]
Oh 🤦 my bad... I realize now that I was getting that error because platform was overriden in the composer.json file. Thanks @Ocramius!!

**Pull request merged**